File: backend/app/api/v1/spending.py
"""What meals paid for cost: Uber Eats and other deliveries, receipts."""

from __future__ import annotations

from datetime import date
from typing import Annotated, Any

from fastapi import APIRouter, Query

from app.core.deps import ReaderDep, SessionDep
from app.core.errors import InvalidInputError
from app.models.base import utcnow
from app.services import spending
from app.services.timed_entries import local_day

router = APIRouter(prefix="/spending", tags=["work-file"])


@router.get("/meals")
async def meals(
    principal: ReaderDep,
    session: SessionDep,
    start: Annotated[date | None, Query()] = None,
    end: Annotated[date | None, Query()] = None,
    kind: Annotated[str, Query(pattern="^(livraison|repas|tout)$")] = (
        "livraison"
    ),
    bucket: Annotated[str, Query(pattern="^(auto|day|week|month)$")] = "auto",
) -> dict[str, Any]:
    """Spending on meals paid for over ``start``..``end``.

    ``kind``: ``livraison`` (Uber Eats and other deliveries, default),
    ``repas`` (receipts, restaurants) or ``tout``. Answers the total, the
    orders, the average basket, the largest order, the monthly average,
    the late orders (21:00 – 05:00), the spending per ``bucket`` (day,
    week, month; ``auto`` from the span, empty periods included), per
    hour of the day, and the establishments that cost the most. Without
    ``start``: from the first order; without ``end``: up to today.
    """
    user_id = principal.user.id
    last = end or await local_day(session, user_id, utcnow())
    if start is not None and start > last:
        raise InvalidInputError("start is after end")
    return await spending.meals(session, user_id, (start, last), kind, bucket)

File: frontend/src/api/spending.ts
// What meals paid for cost: Uber Eats and other deliveries, receipts.
import type { Range } from '../utils/range';
import { api } from './client';

export type SpendKind = 'livraison' | 'repas' | 'tout';
export type Bucket = 'auto' | 'day' | 'week' | 'month';

export interface SpendPeriod {
  /** First day of the day, week or month. */
  start: string;
  total: number;
  count: number;
}

export interface SpendHour {
  hour: number;
  count: number;
  total: number;
  /** Ordered from 21:00 to 04:59. */
  late: boolean;
}

export interface SpendVendor {
  name: string;
  count: number;
  total: number;
  /** Percent of the period's total. */
  share: number;
}

export interface MealSpending {
  start: string;
  end: string;
  bucket: Exclude<Bucket, 'auto'>;
  total: number;
  count: number;
  average: number | null;
  per_month: number | null;
  largest: { amount: number; name: string; at: string } | null;
  late: { count: number; total: number };
  periods: SpendPeriod[];
  hours: SpendHour[];
  vendors: SpendVendor[];
}

export function fetchMealSpending(
  range: Range,
  kind: SpendKind,
  bucket: Bucket,
): Promise<MealSpending> {
  const params = new URLSearchParams({ end: range.end, kind, bucket });
  if (range.start) params.set('start', range.start);
  return api<MealSpending>(`/spending/meals?${params.toString()}`);
}
